presto-spi/src/main/java/com/facebook/presto/spi/PageBuilder.java

```diff
@@ -43,17 +43,27 @@ public PageBuilder(List<? extends Type> types)
 
     public PageBuilder(int initialExpectedEntries, List<? extends Type> types)
     {
+        this(initialExpectedEntries, DEFAULT_MAX_PAGE_SIZE_IN_BYTES, types);
+    }
+
+    public static PageBuilder withMaxPageSize(int maxPageBytes, List<? extends Type> types)
+    {
+        return new PageBuilder(Integer.MAX_VALUE, maxPageBytes, types);
+    }
+
+    private PageBuilder(int initialExpectedEntries, int maxPageBytes, List<? extends Type> types)
```

Comment:
In particular
* setting max block size to `(int) (1.0 * maxPageBytes / types.size())` defeats the purpose of `maxPageSize` in `PageBuilderStatus`
* using max of 32 (when a variable width type exists) and fixed width size, instead of sum, doesn't seem to make sense.
* the special handling of `types.isEmpty()` is far from straightforward.